src/Workspaces/Core/Portable/Utilities/AsyncLazy`1.cs

```diff
@@ -508,13 +505,30 @@ private void OnAsynchronousRequestCancelled(object state)
             }
         }
 
-        // Using inheritance instead of wrapping a TaskCompletionSource to avoid a second allocation
-        private class Request : TaskCompletionSource<T>
+        private sealed class Request
         {
+            /// <summary>
+            /// The <see cref="CancellationToken"/> associated with this request. This field will be initialized before
+            /// any cancellation is observed from the token.
+            /// </summary>
+            private CancellationToken _cancellationToken;
             private CancellationTokenRegistration _cancellationTokenRegistration;
 
+            // We use a AsyncTaskMethodBuilder so we have the ability to cancel the task with a given cancellation token
+            // TODO: remove this once we're on .NET 4.6 and can move back to using TaskCompletionSource
+            private AsyncTaskMethodBuilder<T> _taskBuilder;
+
+            public Task<T> Task
```

Comment:
You should add an explicit constructor to Request that does:
     var ignored = _taskBuilder.Task;
AsyncTaskMethodBuilder is the type underlying async methods, and is tuned for that specific case.  Since a synchronously completing async method doesn't need a non-complete Task allocated, and since the compiler can ensure that before yielding for the first time the builder's Task property is accessed (at a point when no one else has access to it), there is no synchronization involved in creating the task that backs the task builder.  As such, you need to be sure that the Task has been created and stored in the builder's reference before there's any chance of concurrent access to _taskBuilder.